base/strings/string.jl

```diff
@@ -113,28 +113,29 @@ pointer(s::String, i::Integer) = pointer(s) + Int(i)::Int - 1
 ncodeunits(s::String) = Core.sizeof(s)
 codeunit(s::String) = UInt8
 
-@inline function codeunit(s::String, i::Integer)
+@assume_effects :foldable @inline function codeunit(s::String, i::Integer)
     @boundscheck checkbounds(s, i)
     b = GC.@preserve s unsafe_load(pointer(s, i))
     return b
 end
 
 ## comparison ##
 
-_memcmp(a::Union{Ptr{UInt8},AbstractString}, b::Union{Ptr{UInt8},AbstractString}, len) =
+@assume_effects :total function _memcmp(a::Union{Ptr{UInt8},AbstractString}, b::Union{Ptr{UInt8},AbstractString})
```

Comment:
No, this is only safe if both of the arguments are `String`.

Reply:
Or `SubString{String}` I suppose.

Reply:
it's not safe for `SubString` since the compiler can hallucinate out of bounds index for the substring.

Reply:
Good point.